File: lint-arwaky/archive/taxonomy/lint_domain_vo.py
from pydantic import BaseModel, ConfigDict, field_validator, model_validator, Field
from .lint_position_vo import LineNumber, ColumnNumber
from .file_path_vo import FilePath
from .path_collection_vo import PatternList


class ScopeRef(BaseModel):
    """Code scope reference (function, class, module)."""

    model_config = ConfigDict(frozen=True)
    name: str
    kind: str = "function"
    file: FilePath | str = ""
    start_line: LineNumber | int = 0
    end_line: LineNumber | int = 0

    @model_validator(mode="before")
    @classmethod
    def coerce_fields(cls, data: object) -> object:
        if isinstance(data, dict):
            if "file" in data and isinstance(data["file"], str):
                data["file"] = FilePath(value=data["file"])
            if "start_line" in data and isinstance(data["start_line"], int):
                data["start_line"] = LineNumber(value=data["start_line"])
            if "end_line" in data and isinstance(data["end_line"], int):
                data["end_line"] = LineNumber(value=data["end_line"])
        return data

    @field_validator("name")
    @classmethod
    def check_non_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Scope name cannot be empty")
        return v.strip()

    def __str__(self) -> str:
        if self.file:
            prefix = f"{self.kind} " if self.kind else ""
            return f"{prefix}{self.name} in {self.file}"
        if self.kind:
            return f"{self.kind} {self.name}"
        return self.name

    @property
    def has_range(self) -> bool:
        return int(self.start_line) > 0 and int(self.end_line) > 0


class Location(BaseModel):
    """File location with optional description."""

    model_config = ConfigDict(frozen=True)
    file: FilePath | str = ""
    line: LineNumber = Field(default_factory=lambda: LineNumber(value=0))
    column: ColumnNumber = Field(default_factory=lambda: ColumnNumber(value=0))
    description: str = ""

    @model_validator(mode="before")
    @classmethod
    def coerce_from_primitive(cls, data: object) -> object:
        if isinstance(data, dict):
            if "file" in data and isinstance(data["file"], str) and data["file"].strip():
                data["file"] = FilePath(value=data["file"])
            if "line" in data and isinstance(data["line"], int):
                data["line"] = LineNumber(value=data["line"])
            if "column" in data and isinstance(data["column"], int):
                data["column"] = ColumnNumber(value=data["column"])
        return data

    def __str__(self) -> str:
        parts = []
        if str(self.file):
            parts.append(str(self.file))
        if int(self.line) > 0:
            parts.append(str(self.line))
            if int(self.column) > 0:
                parts[-1] += f":{self.column}"
        result = ":".join(parts) if parts else "unknown"
        if self.description:
            result += f" — {self.description}"
        return result


class LocationList(BaseModel):
    """List of Location objects."""

    model_config = ConfigDict(frozen=False)
    values: list[Location] = Field(default_factory=list)

    def __iter__(self):
        return iter(self.values)

    def __len__(self):
        return len(self.values)

    def append(self, item: Location) -> None:
        self.values.append(item)

    def extend(self, other: "LocationList"):
        self.values.extend(other.values)


class ViolationConstraint(BaseModel):
    """Validation constraint descriptor."""

    model_config = ConfigDict(frozen=True)
    rule: str
    min_value: str = ""
    max_value: str = ""

    def __str__(self) -> str:
        parts = [self.rule]
        if self.min_value or self.max_value:
            range_str = (
                f"{self.min_value}..{self.max_value}"
                if self.min_value and self.max_value
                else self.min_value or self.max_value
            )
            parts.append(f"(must be {range_str})")
        return " ".join(parts)


class CommandArgs(BaseModel):
    """Command line arguments list."""

    model_config = ConfigDict(frozen=True)
    args: PatternList = Field(default_factory=PatternList)

    @field_validator("args", mode="before")
    @classmethod
    def validate_args(cls, v):
        if isinstance(v, list):
            return PatternList(values=v)
        return v

    def __str__(self) -> str:
        return " ".join(self.args.values or [])

    def __len__(self) -> int:
        return len(self.args.values or [])


class ScopeBounds(BaseModel):
    """Start and end line numbers of a scope."""

    model_config = ConfigDict(frozen=True)
    start: LineNumber | None = None
    end: LineNumber | None = None

File: lint-arwaky/src-rust/taxonomy/lint_domain_vo.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ScopeRef {
    pub name: String,
    #[serde(default)]
    pub kind: String,
    #[serde(default)]
    pub file: Option<FilePath>,
    #[serde(default)]
    pub start_line: Option<LineNumber>,
    #[serde(default)]
    pub end_line: Option<LineNumber>,
}

impl ScopeRef {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into(), kind: "function".to_string(), file: None, start_line: None, end_line: None }
    }
    pub fn has_range(&self) -> bool {
        self.start_line.as_ref().map_or(false, |l| l.value > 0) &&
        self.end_line.as_ref().map_or(false, |l| l.value > 0)
    }
}

impl std::fmt::Display for ScopeRef {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(ref file) = self.file {
            write!(f, "{} {} in {}", self.kind, self.name, file)
        } else if !self.kind.is_empty() {
            write!(f, "{} {}", self.kind, self.name)
        } else {
            write!(f, "{}", self.name)
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Location {
    #[serde(default)]
    pub file: Option<FilePath>,
    #[serde(default)]
    pub line: Option<LineNumber>,
    #[serde(default)]
    pub column: Option<ColumnNumber>,
    #[serde(default)]
    pub description: String,
}

impl Location {
    pub fn new() -> Self {
        Self { file: None, line: None, column: None, description: String::new() }
    }
}

impl std::fmt::Display for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut parts = Vec::new();
        if let Some(ref file) = self.file { parts.push(file.value.clone()); }
        if let Some(ref line) = self.line {
            let mut s = line.value.to_string();
            if let Some(ref col) = self.column {
                if col.value > 0 { s = format!("{}:{}", line.value, col.value); }
            }
            parts.push(s);
        }
        let result = if parts.is_empty() { "unknown".to_string() } else { parts.join(":") };
        if self.description.is_empty() {
            write!(f, "{}", result)
        } else {
            write!(f, "{} — {}", result, self.description)
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LocationList {
    #[serde(default)]
    pub values: Vec<Location>,
}

impl LocationList {
    pub fn new() -> Self {
        Self { values: Vec::new() }
    }
    pub fn push(&mut self, item: Location) { self.values.push(item); }
    pub fn len(&self) -> usize { self.values.len() }
    pub fn is_empty(&self) -> bool { self.values.is_empty() }
}

impl std::ops::Deref for LocationList {
    type Target = Vec<Location>;
    fn deref(&self) -> &Self::Target { &self.values }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ViolationConstraint {
    pub rule: String,
    #[serde(default)]
    pub min_value: String,
    #[serde(default)]
    pub max_value: String,
}

impl ViolationConstraint {
    pub fn new(rule: impl Into<String>) -> Self {
        Self { rule: rule.into(), min_value: String::new(), max_value: String::new() }
    }
}

impl std::fmt::Display for ViolationConstraint {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.rule)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CommandArgs {
    #[serde(default)]
    pub args: Vec<String>,
}

impl CommandArgs {
    pub fn new() -> Self {
        Self { args: Vec::new() }
    }
}

impl std::fmt::Display for CommandArgs {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.args.join(" "))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ScopeBounds {
    #[serde(default)]
    pub start: Option<LineNumber>,
    #[serde(default)]
    pub end: Option<LineNumber>,
}
